### PR TITLE
✨ Stop a moving Citizens NPC when in a conversation

### PR DESCRIPTION
**Pull request by BakaAless**
All is in the title.

https://user-images.githubusercontent.com/44125445/219726123-c7e12e93-d8ea-4d08-9f44-97562531141f.mp4

### CODE REVIEW

File: paper/src/main/java/rocks/gravili/notquests/paper/conversation/ConversationManager.java [hunk 1]
```diff
@@ -39,6 +35,7 @@
 import rocks.gravili.notquests.paper.conversation.interactionhandlers.SendClickableText;
 import rocks.gravili.notquests.paper.conversation.interactionhandlers.ConversationInteractionHandler;
 import rocks.gravili.notquests.paper.managers.data.Category;
+import rocks.gravili.notquests.paper.managers.npc.CitizensNPC;
```
Comment (AlessioGr) [inline paper/src/main/java/rocks/gravili/notquests/paper/conversation/ConversationManager.java:38]: Don't import CitizensNPC here. Everything outside of integration-specific classes needs to be as generic as possible so we can support more npc plugins easier. Instead, maybe add a method to NqNPC which runs during playConversation and implement that inside of CitizensNPC. That way all Citizens-specific logic can stay in one class

File: paper/src/main/java/rocks/gravili/notquests/paper/conversation/ConversationManager.java [hunk 2]
```diff
@@ -20,12 +20,8 @@
 
 import java.io.File;
 import java.io.IOException;
-import java.util.ArrayList;
-import java.util.Arrays;
-import java.util.HashMap;
-import java.util.List;
-import java.util.Locale;
-import java.util.UUID;
+import java.util.*;
```
Comment (AlessioGr) [inline paper/src/main/java/rocks/gravili/notquests/paper/conversation/ConversationManager.java:23]: not sure a wildcard import for the entire java.util class is a good idea here

File: paper/src/main/java/rocks/gravili/notquests/paper/conversation/ConversationManager.java [hunk 3]
```diff
@@ -55,6 +52,9 @@
 import rocks.gravili.notquests.paper.structs.variables.VariableDataType;
 
 public class ConversationManager {
+
+  public static final Map<Integer, List<UUID>> CONVERSATIONS_IN_PROGRESS = new HashMap<>();
```
Comment (AlessioGr) [inline paper/src/main/java/rocks/gravili/notquests/paper/conversation/ConversationManager.java:56]: hmm what this does is not quite clear from the naming (that it stores npc IDs). Also, is it necessary, considering all open conversations are already stored in "openConversations"?

Also make sure not to use static here (or pretty much anywhere in notquests)

File: paper/src/main/java/rocks/gravili/notquests/paper/conversation/ConversationManager.java (hunk 2 above)
Comment (BakaAless) [reply paper/src/main/java/rocks/gravili/notquests/paper/conversation/ConversationManager.java:23]: IntelliJ reduces automatically 6 imports into only one with a wildcard. Same as in the QuestPlayerManager class.

File: paper/src/main/java/rocks/gravili/notquests/paper/conversation/ConversationManager.java (hunk 3 above)
Comment (BakaAless) [reply paper/src/main/java/rocks/gravili/notquests/paper/conversation/ConversationManager.java:56]: Just to easily retrieve all the conversations that involve this NPC: the conversations can have several speakers and it is much more complex to retrieve the one that involves the current NPC.

File: paper/src/main/java/rocks/gravili/notquests/paper/conversation/ConversationManager.java [hunk 4]
```diff
@@ -546,7 +545,7 @@ public void stopConversation(final ConversationPlayer conversationPlayer) {
                 "Skipping stopping conversation, as the conversation you tried to stop is already stopped and some other conversation is running instead.");
         return;
       }
-      if (conversationPlayer.getNpc() != null && conversationPlayer.getNpc() instanceof CitizensNPC) {
+      if (conversationPlayer.getNpc() != null && conversationPlayer.getNpc().isFromCitizens()) {
```
Comment (AlessioGr) [inline paper/src/main/java/rocks/gravili/notquests/paper/conversation/ConversationManager.java:548]: mh even like that I would change it. So basically, I try to keep all integration-specific _logic_ (I think that makes it more clear) separated. Meaning everything SPECIFIC to the citizens plugin would be in CitizensNPC or CitizensManager or something like that. So no Citizens-specific logic inside of ConversationManager.

It needs to be as generic as possible and extracted (e.g. a method in NqNPC which is implemented in CitizensNPC and called here). That way, say we decide to add support for zNPCs or something like that in the future, we can implement the same functionality into zNPCs way easier.

Comment (BakaAless) [reply paper/src/main/java/rocks/gravili/notquests/paper/conversation/ConversationManager.java:548]: In fact, the issue is from the name; if the method is renamed "canMove", it's fine ?

**Comment by AlessioGr** [comment]
Thank you so much for this cool PR again <3

**Pull request merged**